File: frontend/types/post.ts
export type PostStatus = 'draft' | 'published' | 'scheduled'

export interface Post {
  id: string
  title: string
  content: string
  excerpt?: string
  status: PostStatus
  authorId: string
  author?: {
    id: string
    nickname: string
    profileImage?: string
  }
  categoryId?: string
  category?: Category
  tags?: Tag[]
  viewCount: number
  likeCount: number
  commentCount: number
  isLiked?: boolean // 현재 사용자가 좋아요 했는지
  isBookmarked?: boolean // 현재 사용자가 북마크 했는지
  // SEO 관련
  metaTitle?: string
  metaDescription?: string
  metaKeywords?: string
  // 검색 하이라이팅
  highlightedTitle?: string
  highlightedContent?: string
  publishedAt?: string
  scheduledAt?: string
  createdAt: string
  updatedAt: string
}

export interface Category {
  id: string
  name: string
  slug: string
  parentId?: string
  children?: Category[]
  postCount: number
  createdAt: string
  updatedAt: string
}

export interface Tag {
  id: string
  name: string
  slug: string
  postCount: number
  createdAt: string
  updatedAt: string
}

export interface CreatePostRequest {
  title: string
  content: string
  excerpt?: string
  status: PostStatus
  categoryId?: string
  tagIds?: string[]
  scheduledAt?: string
}

export interface UpdatePostRequest {
  title?: string
  content?: string
  excerpt?: string
  status?: PostStatus
  categoryId?: string
  tagIds?: string[]
  scheduledAt?: string
}

export interface PostListParams {
  page?: number
  limit?: number
  categoryId?: string
  tagId?: string
  status?: PostStatus
  search?: string
  cursor?: string
}

export interface Like {
  id: string
  postId: string
  userId: string
  createdAt: string
}

export interface Bookmark {
  id: string
  postId: string
  userId: string
  createdAt: string
}

File: backend/internal/interface/http/dto/post.go
package dto

import "time"

// CreatePostRequest represents the request to create a new post
type CreatePostRequest struct {
	Title           string   `json:"title" binding:"required,min=1,max=255"`
	Slug            string   `json:"slug" binding:"required,min=1,max=255"`
	Content         string   `json:"content" binding:"required"`
	Excerpt         *string  `json:"excerpt,omitempty"`
	FeaturedImage   *string  `json:"featured_image,omitempty"`
	Status          string   `json:"status" binding:"required,oneof=draft published scheduled"`
	PublishedAt     *time.Time `json:"published_at,omitempty"`
	MetaTitle       *string  `json:"meta_title,omitempty"`
	MetaDescription *string  `json:"meta_description,omitempty"`
	MetaKeywords    *string  `json:"meta_keywords,omitempty"`
	CategoryID      *uint    `json:"category_id,omitempty"`
	TagIDs          []uint   `json:"tag_ids,omitempty" binding:"max=10"` // Max 10 tags per post
}

// UpdatePostRequest represents the request to update a post
type UpdatePostRequest struct {
	Title           *string    `json:"title,omitempty" binding:"omitempty,min=1,max=255"`
	Slug            *string    `json:"slug,omitempty" binding:"omitempty,min=1,max=255"`
	Content         *string    `json:"content,omitempty"`
	Excerpt         *string    `json:"excerpt,omitempty"`
	FeaturedImage   *string    `json:"featured_image,omitempty"`
	Status          *string    `json:"status,omitempty" binding:"omitempty,oneof=draft published scheduled"`
	PublishedAt     *time.Time `json:"published_at,omitempty"`
	MetaTitle       *string    `json:"meta_title,omitempty"`
	MetaDescription *string    `json:"meta_description,omitempty"`
	MetaKeywords    *string    `json:"meta_keywords,omitempty"`
	CategoryID      *uint      `json:"category_id,omitempty"`
	TagIDs          []uint     `json:"tag_ids,omitempty" binding:"max=10"`
}

// PostListRequest represents the query parameters for listing posts
type PostListRequest struct {
	Page  int `form:"page" binding:"omitempty,min=1"`
	Limit int `form:"limit" binding:"omitempty,min=1,max=100"`
}

// PostSearchRequest represents the query parameters for searching posts
type PostSearchRequest struct {
	Query string `form:"q" binding:"required,min=1"`
	Page  int    `form:"page" binding:"omitempty,min=1"`
	Limit int    `form:"limit" binding:"omitempty,min=1,max=100"`
}

// PostResponse represents a post in the response
type PostResponse struct {
	ID              uint                  `json:"id"`
	CreatedAt       time.Time             `json:"created_at"`
	UpdatedAt       time.Time             `json:"updated_at"`
	Title           string                `json:"title"`
	Slug            string                `json:"slug"`
	Content         string                `json:"content"`
	Excerpt         string                `json:"excerpt,omitempty"`
	FeaturedImage   *string               `json:"featured_image,omitempty"`
	Status          string                `json:"status"`
	PublishedAt     *time.Time            `json:"published_at,omitempty"`
	ViewCount       int                   `json:"view_count"`
	LikeCount       int                   `json:"like_count"`
	CommentCount    int                   `json:"comment_count"`
	BookmarkCount   int                   `json:"bookmark_count"`
	MetaTitle       *string               `json:"meta_title,omitempty"`
	MetaDescription *string               `json:"meta_description,omitempty"`
	MetaKeywords    *string               `json:"meta_keywords,omitempty"`
	Author          *AuthorResponse       `json:"author,omitempty"`
	Category        *CategoryResponse     `json:"category,omitempty"`
	Tags            []TagResponse         `json:"tags,omitempty"`
	IsLiked         bool                  `json:"is_liked,omitempty"`         // Whether current user liked the post
	IsBookmarked    bool                  `json:"is_bookmarked,omitempty"`    // Whether current user bookmarked the post
}

// PostListResponse represents a paginated list of posts
type PostListResponse struct {
	Posts      []PostResponse     `json:"posts"`
	Pagination PaginationResponse `json:"pagination"`
}

// AuthorResponse represents the author information in post response
type AuthorResponse struct {
	ID        uint   `json:"id"`
	Nickname  string `json:"nickname"`
	AvatarURL *string `json:"avatar_url,omitempty"`
}

// PaginationResponse represents pagination metadata
type PaginationResponse struct {
	Page       int   `json:"page"`
	Limit      int   `json:"limit"`
	Total      int64 `json:"total"`
	TotalPages int   `json:"total_pages"`
}
